### PR TITLE
Improve ACL 2.0 authorization context handling

### PR DESCRIPTION
**Pull request by dingshuangxi888**
This PR improves ACL 2.0 authorization context handling for remoting and gRPC requests while preserving compatibility with existing request shapes.

### CODE REVIEW

**Review by RongtongJin — APPROVED** [review]

**Review by lizhimins — APPROVED** [review]

File: auth/src/main/java/org/apache/rocketmq/auth/authorization/AuthorizationCompatibility.java [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+            switch (request.getCode()) {
+                case RequestCode.HEART_BEAT:
+                    return isProducerHeartbeat(request);
+                case RequestCode.UNREGISTER_CLIENT:
+                    return isProducerUnregister(request);
+                case RequestCode.END_TRANSACTION:
+                case RequestCode.VIEW_MESSAGE_BY_ID:
+                    return isHistoricalTopicAbsent(request);
+                default:
+                    return false;
```
Comment (RockteMQ-AI) [inline auth/src/main/java/org/apache/rocketmq/auth/authorization/AuthorizationCompatibility.java:50]: The `catch (Throwable ignored)` block silently swallows all exceptions. This is a deliberate defensive pattern for backward compatibility, but consider adding a brief comment explaining why all throwables are caught (e.g., 'catch-all to avoid breaking existing request flows when compatibility detection itself fails'). This helps future maintainers understand the intent and not 'fix' it away.

File: auth/src/main/java/org/apache/rocketmq/auth/authorization/AuthorizationEvaluator.java [hunk 2]
```diff
@@ -36,10 +39,37 @@ public AuthorizationEvaluator(AuthConfig authConfig, Supplier<?> metadataService
         this.authorizationStrategy = AuthorizationFactory.getStrategy(authConfig, metadataService);
     }
 
-    public void evaluate(List<AuthorizationContext> contexts) {
+    /**
```
Comment (RockteMQ-AI) [inline auth/src/main/java/org/apache/rocketmq/auth/authorization/AuthorizationEvaluator.java:42]: The behavior change from silently allowing empty contexts to throwing AuthorizationException is a security improvement. However, this could affect existing deployments that rely on the old pass-through behavior. Consider noting this in the PR description or commit message as a breaking change for operators who have ACL enabled but send requests without proper context building.

**Review by RockteMQ-AI — APPROVED** [review]
﻿### Summary
This PR significantly improves ACL 2.0 authorization context handling for both remoting and gRPC protocols while preserving backward compatibility. The changes are well-structured with comprehensive test coverage.

**Key improvements:**
- **Backward compatibility layer** (`AuthorizationCompatibility`): Gracefully handles legacy request shapes that may not have explicit authorization contexts, preventing breakage for existing deployments
- **Input validation**: `requireResource()` and `decodeRequiredBody()` helpers ensure non-null/non-blank values and proper body decoding, closing potential bypass vectors
- **Extended request coverage**: Many new request codes are now properly handled (BATCH_ACK_MESSAGE, QUERY_ASSIGNMENT, SET_MESSAGE_REQUEST_MODE, etc.)
- **Security hardening**: Retry topic validation against consumer group prevents authorization bypass; empty contexts now properly rejected when compatibility doesn't match
- **Comprehensive tests**: Excellent test coverage including edge cases for compatibility matching, null handling, and resource pattern validation

**Overall**: Solid improvement to the authorization system. The code is clean, well-tested, and maintains backward compatibility.

---
*Automated review by github-manager-bot*
